torch/autograd/_functions/tensor.py

```diff
@@ -601,13 +609,93 @@ def forward(self, input):
         return torch.cumsum(input, dim=self.dim)
 
     def backward(self, grad_output):
-        grad_input = torch.cumsum(-grad_output, dim=self.dim)
+        return reverse_cumsum(grad_output, dim=self.dim)
+        
+class Cumprod(Function):
+    
+    def __init__(self, dim):
+        super(Cumprod, self).__init__()
+        self.dim = dim
 
-        end_idx = grad_input.size(self.dim) - 1
-        grad_sum = grad_input.narrow(self.dim, end_idx, 1)
-        grad_input -= grad_sum.expand_as(grad_input)
-        grad_input += grad_output
-        return grad_input
+    def forward(self, input):
+        self.save_for_backward(input)
+        return torch.cumprod(input, dim=self.dim)
+
+    def backward(self, grad_output):
+        '''
+        There are two algorithms to do this, the naive one
+        which works with all types of inputs is slow, namely
+        O(n^2) where n = input.size(self.dim).
+
+        The second one is much faster, namely O(n), but
+        requires nonzero inputs. This implementation checks
+        for 0s in the inputs and then decides which one to use.
+        '''
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        
+        if dim_size == 1:
+            return grad_output.clone()
+
+        if input.is_cuda:
+            LT = torch.cuda.LongTensor
+        else:
+            LT = torch.LongTensor
+
+        if (input == 0).any():
+            ones = input.index_select(self.dim,
```

Comment:
Why are you using `index_select` with a single element instead of `select` here? If you use select, you don't need to create a LongTensor every time.
If the reason is to keep a dimension of size 1, you can use `narrow` with a length of 1 to avoid creating all the LongTensors.